state/device_data_table.go

```diff
@@ -97,55 +98,90 @@ func (t *DeviceDataTable) DeleteDevice(userID, deviceID string) error {
 // Upsert combines what is in the database for this user|device with the partial entry `dd`
 func (t *DeviceDataTable) Upsert(dd *internal.DeviceData) (err error) {
 	err = sqlutil.WithTransaction(t.db, func(txn *sqlx.Tx) error {
-		// select what already exists
-		var row DeviceDataRow
-		err = txn.Get(&row, `SELECT data FROM syncv3_device_data WHERE user_id=$1 AND device_id=$2`, dd.UserID, dd.DeviceID)
+		// see if anything already exists. We do this by pulling out the changed bit because we need it to calculate the new bitset,
+		// so it's a useful proxy for "does this data exist for this user/device?"
+		// We use FOR UPDATE to block any potential changes to this JSON structure through the lifetime of the transaction.
+		var changedBits int
+		err = txn.QueryRow(`SELECT data->>'c' FROM syncv3_device_data WHERE user_id=$1 AND device_id=$2 FOR UPDATE`, dd.UserID, dd.DeviceID).Scan(&changedBits)
 		if err != nil && err != sql.ErrNoRows {
-			return err
+			return fmt.Errorf("failed to check if device data exists: %v", err)
```

Comment:
Minor nitpick: Please use `%w` instead of `%v` so things like `errors.Is` etc. can work. (Not that we're using it here atm)
Same for the other `fmt.Errorf`